### DISCUSSION TITLE
Preview updating with Microsoft Update

### DISCUSSION THREAD

**Discussion by TravisEz13** [PowerShell/PowerShell#15510] [Enhancements]
Please read the [Preview updating with Microsoft Update](https://devblogs.microsoft.com/powershell/preview-updating-powershell-7-2-with-microsoft-update/) blog for full details.

The blog gives details on how to opt-in and how we will notify you when the updates are available.

Thanks everyone for testing.  ~~An issue has been discovered, to work around, either install 7.1.3 or run the script below.  We are working to resolve this issue as soon as we can.~~  Thanks for all your help.

Update 6/30/2021:
The issue with the update where 7.1 had to be installed has been resolved and the workaround is no longer needed.

Update 11/1/2021: 
Unfortunately, we found a bug in the update produced for 7.2.0-rc.1.  It won't affect any other scenario, but we cannot fix it without delaying 7.2.0.  Thanks for your understanding.

### Update 11/5/2021
When the 7.2.0 update is available, if you have opt'ed in to updating and have MU enabled (or some other source of updates) the update will update 7.1 installs to 7.2.0.  Windows determines the exact date of the update release.  I'll try to update this when I know the exact date.

The script to opt-in to the clean install changes slightly for the non-preview package:

```powershell
$pwshRegPath = "HKLM:\SOFTWARE\Microsoft\PowerShellCore"
$stablePath = Join-Path -Path $pwshRegPath -ChildPath "InstalledVersions\31ab5147-9a97-4452-8443-d9709f0516e1"
if (!(Test-Path -Path $stablePath)) {
    $null = New-Item -Path $stablePath -ItemType Directory -Force
}

Set-ItemProperty -Path $pwshRegPath -Name UseMU -Value 1 -Type DWord
Set-ItemProperty -Path $stablePath -Name Install -Value 1 -Type DWord
```
### Update 11/21/2021

Create discussion dedicated for Microsoft Update release dates.:
https://github.com/PowerShell/PowerShell/discussions/16459

## work around script

~~$pwshRegPath = "HKLM:\SOFTWARE\Microsoft\PowerShellCore"
$stablePath = Join-Path -Path $pwshRegPath -ChildPath "InstalledVersions\31ab5147-9a97-4452-8443-d9709f0516e1"
if (!(Test-Path -Path $stablePath)) {
    $null = New-Item -Path $stablePath -ItemType Directory -Force
}~~

**Comment by kasini3000** [comment]
1 which directory  is powershell preview installed to byMU?
2 Can powershell preview be upgraded silently byMU?

**Reply by TravisEz13** [reply]
1. The update by MU only supports upgrades or installs when it is installed to the default directory.
   - default directory for preview: `%ProgramFiles%\powershell\7-preview`
2. MU only supports silent installs.

**Comment by MaximoTrinidad** [comment]
I reported this issue I'm experiencing after doing the WU update process from Preview.6 to 7: https://github.com/PowerShell/PowerShell/issues/15673
But, I've been experiencing this condition since Preview.6

**Comment by A55imilat0r** [comment]
It's been a week and I don't see it. Will this also target devices on Insider Dev ring?

**Reply by A55imilat0r** [reply]
Additional info is that, I'm in South East Australia region,
On Win 10 Dev Insider ring
I have set the reg key via the script and confirmed the key is in place with the correct value via regedit 
Also have other Microsoft products set for updates in settings

**Reply by TravisEz13** [reply]
I've updated the [main discussion](https://github.com/PowerShell/PowerShell/discussions/15510#discussion-3393910) with a workaround for a known issue.  Can you please give this a try?

**Reply by A55imilat0r** [reply]
Interestingly the issue followed me over to Win 11 insider yesterday, but based on the reg workaround you supplied I can see why. 

The workaround worked instantly. 

Cheers for that :)

**Comment by stknohg** [comment]
For testing, I had tried it on only PowerShell 7.2.0-preview.5 installed environment but I never got an update.
However, I was able to get the update immediately after additional installing PowerShell 7.1.3 stable.

Is it intended behavior?

**Reply by TravisEz13** [reply]
Thanks for finding this issue.

I've updated the [main discussion](https://github.com/PowerShell/PowerShell/discussions/15510#discussion-3393910) with a workaround for a known issue.  Can you please give this a try?

**Reply by stknohg** [reply]
Thanks! This workaround works correctly.

**Comment by kasini3000** [comment]
For testing, I had tried it on only PowerShell 7.2.0-preview.5 installed environment but I never got an update.
i set regedit key already, i'm "zh-cn" user, i install 7.2.0-preview.7 manual today.

**Comment by ghost** [comment]
I set the registry entries in advance for fresh install (I don't currently have PowerShell Preview installed) and have been occasionally checking MS Update. I saw that the Twitter announcement has gone out and still nothing.

See below, registry set, manually rechecking MS Update and seeing nothing, no `pwsh-preview`.
![msups](https://user-images.githubusercontent.com/2061796/123689256-05566300-d821-11eb-91f2-85ed5f185611.png)

**Reply by ghost** [reply]
Can open issue if you want, but this doesn't seem to fit very well in the bug report template?

**Reply by TravisEz13** [reply]
Can you file an issue and include the output of 
```powershell
(Get-CimInstance Win32_OperatingSystem) | Select-Object -Property Version, Caption
```
and run  `Get-WindowsUpdateLog` it will give you a file name, I need the file.

**Reply by TravisEz13** [reply]
I've updated the [main discussion](https://github.com/PowerShell/PowerShell/discussions/15510#discussion-3393910) with a workaround for a known issue.  Can you please give this a try?

**Reply by ghost** [reply]
I have 7.1.3 already installed, and I have run:

```
PS> New-Item -Path HKLM:\SOFTWARE\Microsoft\PowerShellCore\InstalledVersions\31ab5147-9a97-4452-8443-d9709f0516e1 -ItemType Directory -Force
```

After several minutes this has not affected the results in Windows Update. I have submitted #15692 following your other requested steps, in case they are still useful. When possible, I will try a reboot, check updates again, and report back.

**Reply by ghost** [reply]
@TravisEz13
I have rebooted after doing as stated above and unfortunately Windows Update still does not show any PowerShell-related options, and PowerShell Preview has not been installed. I can conduct additional troubleshooting as needed tomorrow.

**Reply by ghost** [reply]
This has been explained for this specific case by organizational policy in Endpoint Manager that defers non-security updates. For that reason, I don't believe I'm able to be helpful for testing of this functionality.

**Reply by ghost** [reply]
I can now report back to say that PowerShell 7.2.0-preview.7 installed on my machine today via Microsoft Update.

**Comment by TravisEz13** [comment]
The issue with the update where 7.1 had to be installed has been resolved and the workaround is no longer needed.

Thanks again for everyone that reported the issue and @stknohg  for finding the work around which pointed to the problem in Microsoft Update.

**Reply by A55imilat0r** [reply]
@TravisEz13 does the work around need to be reversed? I've not seen preview 8 come down from WU yet.

**Reply by TravisEz13** [reply]
You don't have to reverse the work around but you can.

The preview.8 update should have become available on Tuesday. There will be a significant delay from the release on GitHub and when the update become available on MU as we must create the update after we have built the product and work through the process of getting an update available on MU.

**Comment by Jaykul** [comment]
For what it's worth, it worked perfectly just now on a Windows 10 (10.0.19043) system (that already had PS 7.1.3) just now -- except that it installed 7.2.0-preview.7 not preview.8 ... was this a one-time thing?

**Reply by A55imilat0r** [reply]
I think there is a week delay on windows update, as I've not had preview 8 come down yet

**Reply by TravisEz13** [reply]
The preview.8 update should have become available on Tuesday.  There will be a significant delay from the release on GitHub and when the update become available on MU as we must create the update after we have built the product and work through the process of getting an update available on MU.

**Reply by kilasuit** [reply]
preview8 installed for me on my devices yesterday so that tallies with what you've said @TravisEz13

**Comment by kilasuit** [comment]
@TravisEz13 is there a similar registry key for installing the stable version or should this just happen when 7.2.0 goes GA

**Reply by TravisEz13** [reply]
There will be a delay after 7.2 goes GA because we have to on-board each new version we release.

There is an issue here with the documentation about the feature:  https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7787 
Currently, only one registry is needed to opt-in to updates for all versions and the MSI adds it for you.

**Reply by TravisEz13** [reply]
Additionally, the 7.2.0 update will only work in the install scenario, because it won't have an MSI with all the features needed to update.  7.2.1 will be the first update to actually update the stable MSIs.

**Reply by TravisEz13** [reply]
@kilasuit Actually, I had to make last minute fixes to the 7.2.0 update and I change the logic.  I have added an `Update 11/5/2021` to the description of this discussion, with instructions on how to opt-in to both the update and the install scenario.  Update will only be support from 7.1 to reduce the test matrix.

**Reply by kilasuit** [reply]
To clariify my ask the scenarios I was thinking of are

1.  I have a new os install and as part of setting up that os I can make sure I have the right Reg Keys in place and it will install both the stable & preview versions of PowerShell via Windows Update.
2. I have a new os install and as part of setting up that os I can make sure I have the right Reg Keys in place and it will install only the latest stable version of PowerShell via Windows Update.

**Reply by TravisEz13** [reply]
@kilasuit for clean install, as you seem to be asking, this has been asked and answered.  Please read the update in the description of the discussion from 11/5/2021 or clarify what you are asking for.

As for updating, as opposed to clean install, is controlled by a single registry key for both preview and stable.

**Comment by CSBlazingDragon** [comment]
@TravisEz13

This might not be the right area first time using github for posting. 
I am trying to install Powershell 7.2.0 in our test environment to use only WSUS/SCCM disabling Microsoft updates. 

Using the 2 new Parameters to the msi

msiexec.exe /I PowerShell-7.2.0-win-x64.msi "ADD_EXPLORER_CONTEXT_MENU_OPENPOWERSHELL=1", "ENABLE_PSREMOTING=1", "REGISTER_MANIFEST=1", "ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL=1", "USE_MU=1", "ENABLE_MU=0"

However, when I do this I get the attached picture.
![Install_7 2_Update_Params](https://user-images.githubusercontent.com/93357816/141211649-02186b4a-2741-49f5-b6b5-020cd9c58bc0.PNG)

It acts like it is disabling MU as the warning pops into play but the installer still shows it as checked. 
I tried to dig around in the registry and no matter what option I chose I do not see a key that was created either way for EnableMU.
There is one created for USE_MU only if it is enabled, if enabled creates and sets the key to 1. 

Installing manually I don't see anything created for EnableMU, I guess ultimately I just want to confirm it will not use MU and only our SCCM/WSUS Setups. Is there a way to do this in the registry?



Thoughts?

**Reply by TravisEz13** [reply]
Correct, EnableMU isn't a registry key, it's implemented in COM.
 
EnableMU =0 does **not** disable MU.  EnableMU=1 enables it.  You must disable it yourself.  So, it depends on your original state.  This is the behavior we are asked to have by Windows.

There is a DSC resource to configure MU.
Here is an example on how to check in MU is enable programmatically from the resource (this is a little complex, concept):  https://github.com/dsccommunity/xWindowsUpdate/blob/6aab96e94e4a53b817e9f9b7cc403b8036c8b93f/source/DSCResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L596

If you want a feature to enable MU via registry, you must ask Windows.

**Reply by TravisEz13** [reply]
I am thinking about producing a module to do these actions (set Use_MU, setup the clean install registry, and enable MU).  I we could consider also adding the feature to disable MU in that module.  I would not be something quite as complete about dealing with WUA as the DSC Resource, but it would cover the PowerShell Install Scenarios.

Here are the scenarios.
1. Get the state for a package (PowerShell-Preview vs PowerShell) 
2. Set the state for a package.
    * (optionally) Opt-in to Update for the packages
    * (optionally) Opt-in to having MU/WU/WSUS install the package if it's not already there
    * (optionally) Opt-in to MU
    * (very explicitly) opt-out of MU

**Comment by ClientAdmin** [comment]
Hi all
There is still no product category in SCCM Software Update Point for stable PowerShell releases. Is that right or are we missing anything? Please help. Thank you.
Kind regards

**Reply by kilasuit** [reply]
From recent conversations I've seen I believe that we are to expect this to be available until early-mid January

**Comment by EmeraldFlame** [comment]
Looking in my WSUS environment, it doesn't look it's currently pulling in these new powershell updates.

I currently have both the "PowerShell" and "PowerShell Preview - x64" products selected. Are we still waiting for a stable channel to be pushed to WSUS, or is there any guidance on possibly which classifications we need to enable?

**Reply by kilasuit** [reply]
Still waiting on this from what I know am afriad

**Reply by thlac** [reply]
The "Powershell - x64" product listing is now present on our WSUS environment with the 7.2.1 update available

**Reply by kilasuit** [reply]
Seems this installed for me on the 11th and I had not yet noticed 

![image](https://user-images.githubusercontent.com/6355225/153889961-861de5e4-e0ed-4b23-9e9c-f712ad050c43.png)

**Reply by EmeraldFlame** [reply]
Just wanted to come back and follow up that the new "Powershell - x64" product listing is now showing up in my own WSUS environments as well.

**Reply by ClientAdmin** [reply]
I see the product now... but why only the x64 version?

**Comment by pomu0326** [comment]
- [ ]

**Reply by MY-DONNAIZUDDIN** [reply]
Okay

**Comment by MY-DONNAIZUDDIN** [comment]
Its DONED
